scipy/stats/_axis_nan_policy.py

```diff
@@ -0,0 +1,347 @@
+    - ``omit``: NaNs will be omitted when performing the calculation.
+      If insufficient data remains in the axis slice along which the
+      statistic is computed, the corresponding entry of the output will be
+      NaN.
+    - ``raise``: if a NaN is present, a ``ValueError`` will be raised."""
+    .split('\n'))
+_nan_policy_parameter_doc = Parameter(_name, _type, _desc)
+_nan_policy_parameter = inspect.Parameter(_name,
+                                          inspect.Parameter.KEYWORD_ONLY,
+                                          default='propagate')
+
+
+def _axis_nan_policy_factory(result_object, default_axis=0,
+                             n_samples=1, paired=False,
+                             result_unpacker=None, too_small=0):
+    """Factory for a wrapper that adds axis/nan_policy params to a function.
+
+    Parameters
+    ----------
+    result_object : callable
+        Callable that returns an object of the type returned by the function
+        being wrapped (e.g. the namedtuple or dataclass returned by a
+        statistical test) provided the separate components (e.g. statistic,
+        pvalue).
+    default_axis : int, default: 0
+        The default value of the axis argument. Standard is 0 except when
+        backwards compatibility demands otherwise (e.g. `None`).
+    n_samples : int or callable, default: 1
+        The number of data samples accepted by the function
+        (e.g. `mannwhitneyu`), a callable that accepts a dictionary of
+        parameters passed into the function and returns the number of data
+        samples (e.g. `wilcoxon`), or `None` to indicate an arbitrary number
+        of samples (e.g. `kruskal`).
+    paired : {False, True}
+        Whether the function being wrapped treats the samples as paired (i.e.
+        corresponding elements of each sample should be considered as different
+        components of the same sample.)
+    result_unpacker : callable, optional
+        Function that unpacks the results of the function being wrapped into
+        a tuple. This is essentially the inverse of `result_object`. Default
+        is `None`, which is appropriate for statistical tests that return a
+        statistic, pvalue tuple (rather than, e.g., a non-iterable datalass).
+    too_small : int, default: 0
+        The largest unnacceptably small sample for the function being wrapped.
+        For example, some functions require samples of size two or more or they
+        raise an error. This argument prevents the error from being raised when
+        input is not 1D and instead places a NaN in the corresponding element
+        of the result.
+    """
+
+    if result_unpacker is None:
+        def result_unpacker(res):
+            return res[..., 0], res[..., 1]
+
+    def is_too_small(samples):
+        for sample in samples:
+            if len(sample) <= too_small:
+                return True
+        return False
+
+    def axis_nan_policy_decorator(hypotest_fun_in):
```

Comment:
Didn't rename all appearances of `hypotest` yet. Currently, this _is_ only applied to hypothesis tests, so it's not wrong.